### ISSUE TITLE
FTL crash random

### ISSUE THREAD

**Issue by JeroenFree** [pi-hole/FTL#843]
PIhole-FTL  5.1 crash random on  x86 ubuntu machines.

```
[2020-07-18 13:30:48.867 13382/F2136] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-18 13:30:48.867 13382/F2136] ---------------------------->  FTL crashed!  <----------------------------
[2020-07-18 13:30:48.867 13382/F2136] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-18 13:30:48.867 13382/F2136] Please report a bug at https://github.com/pi-hole/FTL/issues
[2020-07-18 13:30:48.867 13382/F2136] and include in your report already the following details:
[2020-07-18 13:30:48.867 13382/F2136] FTL has been running for 16363 seconds
[2020-07-18 13:30:48.867 13382/F2136] FTL branch: master
[2020-07-18 13:30:48.867 13382/F2136] FTL version: v5.1
[2020-07-18 13:30:48.867 13382/F2136] FTL commit: b522646
[2020-07-18 13:30:48.867 13382/F2136] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-18 13:30:48.868 13382/F2136] FTL user: started as pihole, ended as pihole
[2020-07-18 13:30:48.868 13382/F2136] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-18 13:30:48.868 13382/F2136] Process details: MID: 2136
[2020-07-18 13:30:48.868 13382/F2136]                  PID: 13382
[2020-07-18 13:30:48.868 13382/F2136]                  TID: 13382
[2020-07-18 13:30:48.868 13382/F2136]                  Name: pihole-FTL
[2020-07-18 13:30:48.868 13382/F2136] Received signal: Segmentation fault
[2020-07-18 13:30:48.868 13382/F2136]      at address: 0x7f6860b1ffc0
[2020-07-18 13:30:48.868 13382/F2136]      with code:  SEGV_MAPERR (Address not mapped to object)
[2020-07-18 13:30:48.868 13382/F2136] Backtrace:
[2020-07-18 13:30:48.869 13382/F2136] B[0000]: /usr/bin/pihole-FTL(+0x342af) [0x5642a32e62af]
[2020-07-18 13:30:48.884 13382/F2136] L[0000]: /root/project/src/signals.c:119
[2020-07-18 13:30:48.884 13382/F2136] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x128a0) [0x7f68605728a0]
[2020-07-18 13:30:48.885 13382/F2136] B[0002]: /lib/x86_64-linux-gnu/libc.so.6(+0xb1724) [0x7f6860220724]
[2020-07-18 13:30:48.885 13382/F2136] B[0003]: /usr/bin/pihole-FTL(+0xe487d) [0x5642a339687d]
[2020-07-18 13:30:48.913 13382/F2136] L[0003]: /root/project/src/database/sqlite3.c:76531
[2020-07-18 13:30:48.915 13382/F2136] B[0004]: /usr/bin/pihole-FTL(+0xe4c8a) [0x5642a3396c8a]
[2020-07-18 13:30:48.944 13382/F2136] L[0004]: /root/project/src/database/sqlite3.c:83904
[2020-07-18 13:30:48.946 13382/F2136] B[0005]: /usr/bin/pihole-FTL(in_whitelist+0xb8) [0x5642a32ee968]
[2020-07-18 13:30:48.950 13382/F2136] L[0005]: /root/project/src/database/gravity-db.c:704
[2020-07-18 13:30:48.951 13382/F2136] B[0006]: /usr/bin/pihole-FTL(+0x2af89) [0x5642a32dcf89]
[2020-07-18 13:30:48.955 13382/F2136] L[0006]: /root/project/src/dnsmasq_interface.c:261
[2020-07-18 13:30:48.955 13382/F2136] B[0007]: /usr/bin/pihole-FTL(_FTL_new_query+0x69b) [0x5642a32de48b]
[2020-07-18 13:30:48.959 13382/F2136] L[0007]: /root/project/src/dnsmasq_interface.c:435
[2020-07-18 13:30:48.960 13382/F2136] B[0008]: /usr/bin/pihole-FTL(tcp_request+0x758) [0x5642a330e4b8]
[2020-07-18 13:30:48.964 13382/F2136] L[0008]: /root/project/src/dnsmasq/forward.c:1945
[2020-07-18 13:30:48.965 13382/F2136] B[0009]: /usr/bin/pihole-FTL(+0x51d49) [0x5642a3303d49]
[2020-07-18 13:30:48.970 13382/F2136] L[0009]: /root/project/src/dnsmasq/dnsmasq.c:1967
[2020-07-18 13:30:48.970 13382/F2136] B[0010]: /usr/bin/pihole-FTL(main_dnsmasq+0x129c) [0x5642a3305cdc]
[2020-07-18 13:30:48.975 13382/F2136] L[0010]: /root/project/src/dnsmasq/dnsmasq.c:1206
[2020-07-18 13:30:48.975 13382/F2136] B[0011]: /usr/bin/pihole-FTL(main+0xdc) [0x5642a32d925c]
[2020-07-18 13:30:48.979 13382/F2136] L[0011]: /root/project/src/main.c:97
[2020-07-18 13:30:48.979 13382/F2136] B[0012]: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f6860190b97]
[2020-07-18 13:30:48.979 13382/F2136] B[0013]: /usr/bin/pihole-FTL(_start+0x2a) [0x5642a32d93da]
[2020-07-18 13:30:48.984 13382/F2136] L[0013]: ??:?
[2020-07-18 13:30:48.985 13382/F2136] ------ Listing content of directory /dev/shm ------
[2020-07-18 13:30:48.985 13382/F2136] File Mode User:Group  Filesize Filename
[2020-07-18 13:30:48.985 13382/F2136] rwxrwxrwx root:root 260 .
[2020-07-18 13:30:48.985 13382/F2136] rwxr-xr-x root:root 4K ..
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 66K FTL-dns-cache
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 45K FTL-overTime
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 3M FTL-queries
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 332K FTL-clients
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 98K FTL-domains
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 102K FTL-strings
[2020-07-18 13:30:48.985 13382/F2136] rw------- pihole:pihole 12 FTL-settings
[2020-07-18 13:30:48.986 13382/F2136] rw------- pihole:pihole 144 FTL-counters
[2020-07-18 13:30:48.986 13382/F2136] rw------- pihole:pihole 48 FTL-lock
[2020-07-18 13:30:48.986 13382/F2136] ---------------------------------------------------
[2020-07-18 13:30:48.986 13382/F2136] Thank you for helping us to improve our FTL engine!
[2020-07-18 13:30:48.986 13382/F2136] Asking parent pihole-FTL (PID 2136) to shut down
[2020-07-18 13:30:48.986 13382/F2136] FTL fork terminated!
[2020-07-18 13:30:48.986 2136M] Received: Real-time signal 2 (36 -> 2)
[2020-07-18 13:30:48.986 2136M] Shutting down...
[2020-07-18 13:30:48.987 2136M] Finished final database update
[2020-07-18 13:30:48.988 2136M] ########## FTL terminated after 1.636319e+04 s! ##########
```

**Comment by DL6ER** [comment]
@JeroenFree Did this crash appear only once or is it reproducible?

**Comment by DL6ER** [comment]
This may be related to #841

**Comment by JeroenFree** [comment]
Yes if client ask a domain. Wit only a point 
Then hes crashing

Verzonden vanaf mijn Xperia van Sony-smartphone

---- DL6ER schreef ----

**Comment by DL6ER** [comment]
Please try if
```
pihole checkout ftl fix/guard_gravity_reopening
```
fixes the issues you're seeing.

**Comment by JeroenFree** [comment]
looks Better let it run for hrs

**Comment by JeroenFree** [comment]
Nooooo Stil Crash 
[2020-07-24 01:01:24.144 4983/F1514] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-24 01:01:24.144 4983/F1514] ---------------------------->  FTL crashed!  <----------------------------
[2020-07-24 01:01:24.144 4983/F1514] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2020-07-24 01:01:24.144 4983/F1514] Please report a bug at https://github.com/pi-hole/FTL/issues
[2020-07-24 01:01:24.144 4983/F1514] and include in your report already the following details:
[2020-07-24 01:01:24.144 4983/F1514] FTL has been running for 3538 seconds
[2020-07-24 01:01:24.144 4983/F1514] FTL branch: fix/guard_gravity_reopening
[2020-07-24 01:01:24.144 4983/F1514] FTL version: vDev-3af4f87
[2020-07-24 01:01:24.144 4983/F1514] FTL commit: 3af4f87
[2020-07-24 01:01:24.144 4983/F1514] FTL date: 2020-07-20 00:07:01 +0200
[2020-07-24 01:01:24.144 4983/F1514] FTL user: started as pihole, ended as pihole
[2020-07-24 01:01:24.144 4983/F1514] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-24 01:01:24.144 4983/F1514] Process details: MID: 1514
[2020-07-24 01:01:24.144 4983/F1514]                  PID: 4983
[2020-07-24 01:01:24.144 4983/F1514]                  TID: 4983
[2020-07-24 01:01:24.144 4983/F1514]                  Name: pihole-FTL

2020-07-24 01:01:24.144 4983/F1514] Received signal: Segmentation fault
[2020-07-24 01:01:24.144 4983/F1514]      at address: 0x7fbcb2352fc0
[2020-07-24 01:01:24.144 4983/F1514]      with code:  SEGV_MAPERR (Address not mapped to object)
[2020-07-24 01:01:24.145 4983/F1514] Backtrace:
[2020-07-24 01:01:24.145 4983/F1514] B[0000]: /usr/bin/pihole-FTL(+0x3437f) [0x55861bf3737f]
[2020-07-24 01:01:24.161 4983/F1514] L[0000]: /root/project/src/signals.c:119
[2020-07-24 01:01:24.161 4983/F1514] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x128a0) [0x7fbcb1d8e8a0]
[2020-07-24 01:01:24.161 4983/F1514] B[0002]: /lib/x86_64-linux-gnu/libc.so.6(+0xb1724) [0x7fbcb1a3c724]
[2020-07-24 01:01:24.161 4983/F1514] B[0003]: /usr/bin/pihole-FTL(+0xe4b5d) [0x55861bfe7b5d]
[2020-07-24 01:01:24.190 4983/F1514] L[0003]: /root/project/src/database/sqlite3.c:76531
[2020-07-24 01:01:24.192 4983/F1514] B[0004]: /usr/bin/pihole-FTL(+0xe4f6a) [0x55861bfe7f6a]
[2020-07-24 01:01:24.221 4983/F1514] L[0004]: /root/project/src/database/sqlite3.c:83904
[2020-07-24 01:01:24.223 4983/F1514] B[0005]: /usr/bin/pihole-FTL(in_whitelist+0xb8) [0x55861bf3fc48]
[2020-07-24 01:01:24.227 4983/F1514] L[0005]: /root/project/src/database/gravity-db.c:712
[2020-07-24 01:01:24.227 4983/F1514] B[0006]: /usr/bin/pihole-FTL(+0x2aff9) [0x55861bf2dff9]
[2020-07-24 01:01:24.231 4983/F1514] L[0006]: /root/project/src/dnsmasq_interface.c:261
[2020-07-24 01:01:24.232 4983/F1514] B[0007]: /usr/bin/pihole-FTL(_FTL_new_query+0x69b) [0x55861bf2f4fb]
[2020-07-24 01:01:24.236 4983/F1514] L[0007]: /root/project/src/dnsmasq_interface.c:435
[2020-07-24 01:01:24.236 4983/F1514] B[0008]: /usr/bin/pihole-FTL(tcp_request+0x758) [0x55861bf5f798]
[2020-07-24 01:01:24.241 4983/F1514] L[0008]: /root/project/src/dnsmasq/forward.c:1945
[2020-07-24 01:01:24.242 4983/F1514] B[0009]: /usr/bin/pihole-FTL(+0x52029) [0x55861bf55029]
[2020-07-24 01:01:24.246 4983/F1514] L[0009]: /root/project/src/dnsmasq/dnsmasq.c:1967
[2020-07-24 01:01:24.247 4983/F1514] B[0010]: /usr/bin/pihole-FTL(main_dnsmasq+0x129c) [0x55861bf56fbc]
[2020-07-24 01:01:24.252 4983/F1514] L[0010]: /root/project/src/dnsmasq/dnsmasq.c:1206
[2020-07-24 01:01:24.252 4983/F1514] B[0011]: /usr/bin/pihole-FTL(main+0xdc) [0x55861bf2a29c]
[2020-07-24 01:01:24.256 4983/F1514] L[0011]: /root/project/src/main.c:97
[2020-07-24 01:01:24.256 4983/F1514] B[0012]: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7fbcb19acb97]
[2020-07-24 01:01:24.256 4983/F1514] B[0013]: /usr/bin/pihole-FTL(_start+0x2a) [0x55861bf2a41a]
[2020-07-24 01:01:24.261 4983/F1514] L[0013]: ??:?
[2020-07-24 01:01:24.262 4983/F1514] ------ Listing content of directory /dev/shm ------
[2020-07-24 01:01:24.262 4983/F1514] File Mode User:Group  Filesize Filename
[2020-07-24 01:01:24.262 4983/F1514] rwxrwxrwx root:root 260 .
[2020-07-24 01:01:24.262 4983/F1514] rwxr-xr-x root:root 4K ..
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 25K FTL-dns-cache
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 45K FTL-overTime
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 1M FTL-queries
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 332K FTL-clients
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 295K FTL-domains
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 188K FTL-strings
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 12 FTL-settings
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 144 FTL-counters
[2020-07-24 01:01:24.262 4983/F1514] rw------- pihole:pihole 48 FTL-lock
[2020-07-24 01:01:24.262 4983/F1514] ---------------------------------------------------
[2020-07-24 01:01:24.262 4983/F1514] Thank you for helping us to improve our FTL engine!
[2020-07-24 01:01:24.262 4983/F1514] Asking parent pihole-FTL (PID 1514) to shut down
[2020-07-24 01:01:24.263 4983/F1514] FTL fork terminated!
[2020-07-24 01:01:24.263 1514M] Received: Real-time signal 2 (36 -> 2)
[2020-07-24 01:01:24.263 1514M] Shutting down...

**Comment by DL6ER** [comment]
The crash happened in a TCP worker fork.

Please add
```
DEBUG_DATABASE=true
DEBUG_VECTORS=true
```
to the file `/etc/pihole/pihole-FTL.conf` (create if it does not exist) and run
```
pihole restartdns
```

Next time the crash happens, please include some lines above the `!!!!!` line. This may help us getting a better picture of where/why exactly the issue happens.

**Comment by JeroenFree** [comment]
Tomorrow in the morning i add to config

Greets Jeroen 

Verzonden vanaf mijn Xperia van Sony-smartphone

---- DL6ER schreef ----

**Comment by JeroenFree** [comment]
Test on the new 5.2 and running long time :) 
last night at 01:00 hangs it own and pc is offline

**Comment by DL6ER** [comment]
This should be fixed by the release of Pi-hole FTL 5.2. Feel free to re-open of create a new ticket with new information if the issue persists.
